executor/aggfuncs/builder.go

```diff
@@ -155,64 +166,95 @@ func buildAvg(aggFuncDesc *aggregation.AggFuncDesc, ordinal int) AggFunc {
 			return &avgOriginal4Float64{baseAvgFloat64{base}}
 		}
 
-	// Build avg functions which consume the partial result of other avg
-	// functions and update their partial results.
-	case aggregation.Partial2Mode, aggregation.FinalMode:
-		switch aggFuncDesc.Args[1].GetType().Tp {
+		// Build avg functions which consume the partial result of other avg
+		// functions and update their partial results.
+	case aggregation.Partial2Mode:
+		switch aggFuncDesc.RetTp.Tp {
 		case mysql.TypeNewDecimal:
 			return &avgPartial4Decimal{baseAvgDecimal{base}}
 		case mysql.TypeDouble:
 			return &avgPartial4Float64{baseAvgFloat64{base}}
 		}
+	case aggregation.FinalMode:
+		switch aggFuncDesc.RetTp.Tp {
+		case mysql.TypeNewDecimal:
+			return buildFunc4AvgDecimal(finalOrdinal, partialOrdinal, aggFuncDesc.RetTp)
+		case mysql.TypeDouble:
+			return buildFunc4AvgFloat64(finalOrdinal, partialOrdinal, aggFuncDesc.RetTp)
+		}
 	}
-	return nil
+	return
 }
 
 // buildFirstRow builds the AggFunc implementation for function "FIRST_ROW".
-func buildFirstRow(aggFuncDesc *aggregation.AggFuncDesc, ordinal int) AggFunc {
+func buildFirstRow(aggFuncDesc *aggregation.AggFuncDesc, finalOrdinal int, partialOrdinal int) (aggFunc AggFunc) {
 	base := baseAggFunc{
 		args:    aggFuncDesc.Args,
-		ordinal: ordinal,
+		ordinal: []int{finalOrdinal, partialOrdinal},
 	}
 
 	evalType, fieldType := aggFuncDesc.RetTp.EvalType(), aggFuncDesc.RetTp
 	if fieldType.Tp == mysql.TypeBit {
 		evalType = types.ETString
 	}
-	switch aggFuncDesc.Mode {
+	switch mode := aggFuncDesc.Mode; mode {
```

Comment:
Just `switch mode := aggFuncDesc.Mode {}` is okay?

Reply:
noop

Reply:
What error will this cause?

Reply:
syntax error.
you may have a try here.
remove the `os` after  the `;`